---
layout: post
title: 关于socket异常那些事儿
tagline: socket;mina;TCP;CLOSED_WAIT;FIN_WAIT2;openfire
category: null
tags: []
published: true
---
最近发现Openfire服务端隔段时间就会报这样的错：

![image](/assets/post-images/2014-05-05-0309a47f-f953-4757-b58b-027552a6f390.png)

有时候是这样：

![image](/assets/post-images/2014-05-05-135c7f7c-fba6-4988-f536-bfa393f845cb.png)

决定处理一下这个“黑盒”，从异常信息看，这个异常有关socket连接超时或者是连接被重置。查了一下相关问题出现原因：

常出现的Connection reset by peer: 原因可能是多方面的，不过更常见的原因是： 

①：服务器的并发连接数超过了其承载量，服务器会将其中一些连接Down掉； 
②：客户端断掉连接，而服务器还在给客户端发送数据； 

该异常在客户端和服务器端均有可能发生，引起该异常的原因有两个，一个是如果一端的Socket被关闭（或主动关闭或者因为异常退出而引起的关闭），另一端仍发送数据，发送的第一个数据包引发该异常(Connect reset by peer)。另一个是一端退出，但退出时并未关闭该连接，另一端如果在从连接中读数据则抛出该异常（Connection reset）。简单的说就是由连接断开后的读和写操作引起的。

如果频繁出现，就表示很多客户端连接到Apache服务器的响应时间太长了，可能是网络的问题或者服务器性能问题。

Connection timed out: 

连接超时，其实还是找不到已连接的客户端。可能出现原因：客户端网络问题导致连接没断开。

由于Openfire是基于mina做的socket连接，mina本身封装了socket的各种异常，只剩下 ProtocolCodecException、ProtocolEncoderException、ProtocolDecoderException、RecoverableProtocolDecoderException 四种异常，其他都作为运行时异常抛出。

Openfire的连接管理类是ConnectionHandler，继承了mina的IoHandlerAdapter。通过exceptionCaught方法实现连接过程中的异常处理。

可以看到源码如下(Openfire 3.7.2版本)：

    @Override
  	public void exceptionCaught(IoSession session, Throwable cause) throws Exception {
          if (cause instanceof IOException) {
              // TODO Verify if there were packets pending to be sent and decide what to do with them
              Log.info("ConnectionHandler reports IOException for session: " + session, cause);
          }
          else if (cause instanceof ProtocolDecoderException) {
              Log.warn("Closing session due to exception: " + session, cause);
              
              // PIO-524: Determine stream:error message.
              final StreamError error;
              if (cause.getCause() != null && cause.getCause() instanceof XMLNotWellFormedException) {
              	error = new StreamError(StreamError.Condition.xml_not_well_formed);
              } else {
              	error = new StreamError(StreamError.Condition.internal_server_error);
              }
              
              final Connection connection = (Connection) session.getAttribute(CONNECTION);
              connection.deliverRawText(error.toXML());
              session.close();
          }
          else {
              Log.error("ConnectionHandler reports unexpected exception for session: " + session, cause);
          }
      }



